primitives/src/runtime_function.rs

```diff
@@ -0,0 +1,139 @@
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Polkadot runtime functions.
+//! This describes a function that can be called from an external transaction.
+
+use bytes::Vec;
+use codec::Slicable;
+
+/// Public functions that can be dispatched to.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[repr(u8)]
+enum FunctionId {
+	/// Staking subsystem: begin staking.
+	StakingStake = 0,
+	/// Staking subsystem: stop staking.
+	StakingUnstake = 1,
+	/// Staking subsystem: transfer stake.
+	StakingTransfer = 2,
+	/// Set temporary session key as a validator.
+	SessionSetKey = 3,
+	/// Set the timestamp.
+	TimestampSet = 4,
+	/// Make a proposal for the governance system.
+	GovernancePropose = 5,
+	/// Approve a proposal for the governance system.
+	GovernanceApprove = 6,
+}
+
+impl FunctionId {
+	/// Derive `Some` value from a `u8`, or `None` if it's invalid.
+	fn from_u8(value: u8) -> Option<FunctionId> {
+		match value {
+			0 => Some(FunctionId::StakingStake),
+			1 => Some(FunctionId::StakingUnstake),
+			2 => Some(FunctionId::StakingTransfer),
+			3 => Some(FunctionId::SessionSetKey),
+			4 => Some(FunctionId::TimestampSet),
+			5 => Some(FunctionId::GovernancePropose),
+			6 => Some(FunctionId::GovernanceApprove),
+			_ => None,
+		}
+	}
+}
+
+/// Functions on the runtime.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub enum Function {
+	/// Staking subsystem: begin staking.
+	StakingStake,
+	/// Staking subsystem: stop staking.
+	StakingUnstake,
+	/// Staking subsystem: transfer stake.
+	StakingTransfer(::AccountId, ::Balance),
+	/// Set temporary session key as a validator.
+	SessionSetKey(::SessionKey),
+	/// Set the timestamp.
+	TimestampSet(::Timestamp),
+	/// Make a proposal for the governance system.
+	GovernancePropose(::proposal::Proposal),
+	/// Approve a proposal for the governance system.
+	GovernanceApprove(::block::Number),
+}
+
+impl Slicable for Function {
+	fn from_slice(value: &mut &[u8]) -> Option<Self> {
+		let id = try_opt!(u8::from_slice(value).and_then(FunctionId::from_u8));
+		Some(match id {
+			FunctionId::StakingStake => Function::StakingStake,
+			FunctionId::StakingUnstake => Function::StakingUnstake,
+			FunctionId::StakingTransfer => Function::StakingTransfer(
+				try_opt!(Slicable::from_slice(value)),
```

Comment:
does this really deserialise in the correct order?

Reply:
according to eddyb, function argument evaluation order is from left to right: https://www.reddit.com/r/rust/comments/4grfi3/argument_evaluation_order/d2k5887/

but I can change it if desired.